.changeset/chilly-wombats-knock.md

```diff
@@ -0,0 +1,12 @@
+---
+'@backstage/app-defaults': minor
+'@backstage/core-app-api': minor
+'@backstage/create-app': minor
+'@backstage/core-plugin-api': minor
+'@backstage/test-utils': minor
+---
+
+Add support for WebSocket in frontend plugins.
```

Comment:
This is exactly the reason this PR needs discussion, thanks for the comment! For now I haven't come up with any other use-cases than the notifications. Long-polling is always an option for most use cases but I don't think it's the most resource efficient way to do things. I think this boils down to two questions:

1. Does community need support for WebSockets in Backstage?
2. Do WebSockets sit in your vision of the future architecture of Backstage? And if, where should they be used?

Before I continue fixing some of your good points, I think it makes sense to get a clear vision where this should go. Don't want to spend more hours on something that gets thrown away 😁

Reply:
WebSockets are usually used for chats so that might be a use case. Also if some plugin wants to offer real time data for example for monitoring purposes, it might come handy. Not sure if there's already some plugins that might use this..?

Reply:
I feel it's best to approach these not from the point of view of WebSockets, but the higher level capability that they can provide. You never "need" WebSockets, they're a technical implementation that can solve different problems. Now with that said, it's tricky to set up WebSockets unless you have support from the framework with our current architecture. That means there's space for a lower-level capability, but I think it's best not to mix that with higher level concerns. I'd break it down into whether we need these things:

1. A platform-level solution for mid-latency (~1s) backend-to-frontend push
1. A platform-level solution for low-latency (~0.1s) bidirectional frontend/backend communication
1. A platform-level solution for low-latency bidirectional communication between clients
1. A platform-level solution for upgrading WebSocket connections, so that plugins or services can set up their own WebSocket connections

Fill in if I've missed anything 😁

My opinion at this point is that we see a need for 1, and that 4 would be nice to have.

Reply:
Hey @Rugvip!

Forgive my ignorance on this subject, but other than displaying notifications, are there any more examples of point 1 you could give? 😅

I'm just curious. I'm not sure at the moment how or if we'd use something like this, but maybe I'm not fully seeing the potential 🤔

Reply:
This made me realize that would this all make much more sense if it was named differently e.g. EventsApi, EventsServer etc. That way the underlying implementation could be changed from WebSockets to something else. Of course the name collides with the Events backend but it could actually be integrated to that to also allow events to frontend, like @Rugvip mentioned earlier.

Reply:
> This made me realize that would this all make much more sense if it was named differently e.g. EventsApi, EventsServer etc. That way the underlying implementation could be changed from WebSockets to something else. Of course the name collides with the Events backend but it could actually be integrated to that to also allow events to frontend, like @Rugvip mentioned earlier.

This certainly has a better ring to it (at least for me) 💯

Reply:
Refactored it to Events instead Websockets. And yeah, it has better sound to it now, but it collides pretty badly with the events backend naming. Also added support for the events backend to utlize this, so basically any frontend plugin could already subscribe to the events backend events (not sure if events frontend plugin is needed separately).

Have you had any further discussion internally, @Rugvip? Personally I see that Backstage needs some kind of async communication mechanism bi-directionally between frontend and backend like you mentioned in your comment. While it might be possible to start up a websocket server for each plugin individually, I think it might end up being a mess. Also this would not support plugins to interact with each others. 

Let me know your thoughts, thanks again!

Reply:
Gonna resolve this with a quick summary of the way forward:

- Only target the mid-latency server push use-case, pushing signals to web clients from backends
- Re-use events backend where possible, avoid duplication
- Use a transport abstraction that supports fallbacks like SSE or long-polling, but we can do WebSockets too
- Web clients should not be able to subscribe to arbitrary event backend topics, the signals use a different topic namespace

Fill in if I forgot anything 😁